File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Vector2D.java
package org.firstinspires.ftc.teamcode.util;


public class Vector2D {
    public double x,y;

    public Vector2D(double x, double y){
        this.x = x;
        this.y = y;
    }

    public static Vector2D fromHeadingAndMagnitude(double h, double m){
        return new Vector2D(Math.cos(h) * m, Math.sin(h) * m);
    }

    public Vector2D mult(double scalar) {
        return new Vector2D(x * scalar, y * scalar);
    }

    public Vector2D divide(double scalar) {
        return new Vector2D(x / scalar, y / scalar);
    }

    public Vector2D subt(Vector2D other) {
        return new Vector2D(x - other.x, y - other.y);
    }

    public double dot(Vector2D other) {
        return x * other.x + y * other.y;
    }

    public double magnitude() {
        return Math.hypot(x, y);
    }

    public Vector2D unit() {
        return this.divide(magnitude());
    }

    public Vector2D rotate(double angle) {
        return new Vector2D(
                x * Math.cos(angle) - y * Math.sin(angle),
                x * Math.sin(angle) + y * Math.cos(angle));
    }

    public double cross(Vector2D other) {
        return x * other.y - y * other.x;
    }

    @Override
    public String toString() {
        return String.format("{%.2f, %.2f}", x, y);
    }
}

File: RoadRunner/src/main/kotlin/com/acmerobotics/roadrunner/geometry/Vector2d.kt
package com.acmerobotics.roadrunner.geometry

import com.acmerobotics.roadrunner.util.Angle
import com.acmerobotics.roadrunner.util.epsilonEquals
import kotlin.math.*

/**
 * Class for representing 2D vectors (x and y).
 */
data class Vector2d @JvmOverloads constructor(
    val x: Double = 0.0,
    val y: Double = 0.0
) {
    companion object {
        @JvmStatic
        fun polar(r: Double, theta: Double) = Vector2d(r * cos(theta), r * sin(theta))
    }

    fun norm() = sqrt(x * x + y * y)

    fun angle() = Angle.norm(atan2(y, x))

    infix fun angleBetween(other: Vector2d) = acos((this dot other) / (norm() * other.norm()))

    operator fun plus(other: Vector2d) =
        Vector2d(x + other.x, y + other.y)

    operator fun minus(other: Vector2d) =
        Vector2d(x - other.x, y - other.y)

    operator fun times(scalar: Double) = Vector2d(scalar * x, scalar * y)

    operator fun div(scalar: Double) = Vector2d(x / scalar, y / scalar)

    operator fun unaryMinus() = Vector2d(-x, -y)

    infix fun dot(other: Vector2d) = x * other.x + y * other.y

    infix fun distTo(other: Vector2d) = (this - other).norm()

    infix fun projectOnto(other: Vector2d) = other * (this dot other) / (other dot other)

    fun rotated(angle: Double): Vector2d {
        val newX = x * cos(angle) - y * sin(angle)
        val newY = x * sin(angle) + y * cos(angle)
        return Vector2d(newX, newY)
    }

    infix fun epsilonEquals(other: Vector2d) =
        x epsilonEquals other.x && y epsilonEquals other.y

    override fun toString() = String.format("(%.3f, %.3f)", x, y)
}

operator fun Double.times(vector: Vector2d) = vector.times(this)

operator fun Double.div(vector: Vector2d) = vector.div(this)
